balance-reporter/src/index.ts

```diff
@@ -0,0 +1,121 @@
+import { U64 } from "codechain-primitives";
+import Rpc from "codechain-rpc";
+import { createEmail } from "./Email";
+
+type balanceInfo = {
+    address: string;
+    balance: U64;
+};
+
+function getConfig(field: string, defaultVal?: string): string {
+    const c = process.env[field];
+    if (c == null) {
+        if (defaultVal == null) {
+            throw new Error(`${field} is not specified`);
+        }
+        return defaultVal;
+    }
+    return c;
+}
+
+const rpcUrl = getConfig("RPC_URL");
+const sendgridApiKey = getConfig("SENDGRID_API_KEY");
+const sendgridTo = getConfig("SENDGRID_TO");
+const minAllowedBalance = new U64(getConfig("MIN_BALANCE"));
+const maxAllowedBalance = new U64(getConfig("MAX_BALANCE"));
+
+const rpc = new Rpc(rpcUrl);
+const email = createEmail({
+    tag: `[mainnet][balance-reporter]`,
+    sendgridApiKey,
+    to: sendgridTo,
+});
+
+const targetAddresses: string[] = process.argv.slice(2);
+
+function outOfRange(target: U64, min: U64, max: U64) {
+    return target.lt(min) || target.gt(max);
+}
+
+async function getCCCBalances(
+    addresses: string[],
+    blockNumber: number,
+): Promise<Array<balanceInfo>> {
+    const balances = await Promise.all(
+        addresses.map(address => rpc.chain.getBalance({ address, blockNumber })),
+    );
+    const result = [];
+    for (let i = 0; i < addresses.length; i++) {
+        const inst = balances[i];
+        if (inst != null) {
+            result.push({
+                address: addresses[i],
+                balance: new U64(inst),
+            });
+        }
+    }
+    return result;
+}
+
+function toReportMessage(prefix: string, infos: balanceInfo[], blockNumber: number): string {
+    const infosString = infos
+        .map(info => `<li> ${info.address} : ${info.balance} </li>`)
+        .join("<br />\r\n");
+    return `<p>${prefix}, block number: ${blockNumber}</p>
+    <ul>${infosString}</ul>
+    `;
+}
+
+async function getNextBlockNumber(current: number) {
```

Comment:
Considering the purpose of this script, you don't need to see the balances of all blocks.

Reply:
I made this to check all the blocks to report immediately when the balances become out of range.
Then is it better to check at a suitable time interval ?